code/go/0chain.net/smartcontract/minersc/miner.go

```diff
@@ -179,50 +182,36 @@ func (msc *MinerSmartContract) deleteNode(
 ) (*MinerNode, error) {
 	var err error
 	deleteNode.Delete = true
-
-	// deleting pending pools
-	for key, pool := range deleteNode.Pending {
-		var un *UserNode
-		if un, err = msc.getUserNode(pool.DelegateID, balances); err != nil {
-			return nil, fmt.Errorf("getting user node: %v", err)
-		}
-
-		var transfer *state.Transfer
-		transfer, _, err = pool.EmptyPool(msc.ID, pool.DelegateID, nil)
-		if err != nil {
-			return nil, fmt.Errorf("error emptying delegate pool: %v", err)
-		}
-
-		if err = balances.AddTransfer(transfer); err != nil {
-			return nil, fmt.Errorf("adding transfer: %v", err)
-		}
-
-		if err := un.deletePool(deleteNode.ID, key); err != nil {
-			return nil, fmt.Errorf("deleting pool: %v", err)
-		}
-		delete(deleteNode.Pending, key)
-
-		if err = un.save(balances); err != nil {
-			return nil, fmt.Errorf("saving user node%s: %v", un.ID, err)
-		}
+	var nodeType stakepool.Provider
+	if deleteNode.NodeType == NodeTypeMiner {
+		nodeType = stakepool.Miner
+	} else {
+		nodeType = stakepool.Sharder
 	}
 
-	// deleting active pools
-	for key, pool := range deleteNode.Active {
-		if pool.Status == DELETING {
-			continue
-		}
-
-		pool.Status = DELETING // mark as deleting
-		pool.TokenLockInterface = &ViewChangeLock{
-			Owner:               pool.DelegateID,
-			DeleteViewChangeSet: true,
-			DeleteVC:            gn.ViewChange,
+	for key, pool := range deleteNode.Pools {
+		switch pool.Status {
+		case stakepool.Pending:
+			_, err := deleteNode.UnlockPool(
+				pool.DelegateID, nodeType, deleteNode.ID, key, balances)
```

Comment:
The UnlockPool will get the UserStakePool from MPT each time when it is called, we need to have a more efficient way to unlock pools that all belong to the same node here.

Maybe add a new method UnlockPools that pass the pool keys slice and update all at once.